src/main/java/com/nanalysis/sharepoint/SharepointClient.java

```diff
@@ -0,0 +1,251 @@
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.nanalysis.sharepoint;
+
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.w3c.dom.Element;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.CookieManager;
+import java.net.CookiePolicy;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * A sharepoint client to manage authentication, files and folders.
+ */
+public class SharepointClient {
+    private final HttpClient httpClient = HttpClient.newBuilder().cookieHandler(new CookieManager(null, CookiePolicy.ACCEPT_ALL)).build();
+    private final String baseUrl;
+    private final String siteUrl;
+
+    private String token = "";
+
+    public SharepointClient(String baseUrl, String site) {
+        this.baseUrl = baseUrl;
+        this.siteUrl = baseUrl + "/sites/" + site;
+    }
+
+    public void authenticate(String username, String password)
+            throws IOException, InterruptedException, XPathExpressionException, ParserConfigurationException, SAXException {
+        String token = getSecurityToken(username, password);
+        signin(token);
+        this.token = fetchBearerToken();
+    }
+
+    private String getSecurityToken(String username, String password) throws IOException, InterruptedException,
+            ParserConfigurationException, SAXException, XPathExpressionException {
+        String authentication = Templates.REQUEST_SECURITY_TOKEN
+                .replace("${username}", username)
+                .replace("${password}", password)
+                .replace("${endpoint}", baseUrl);
+
+        HttpRequest request = HttpRequest.newBuilder()
+                .uri(URI.create("https://login.microsoftonline.com/extSTS.srf"))
+                .POST(HttpRequest.BodyPublishers.ofString(authentication))
+                .build();
+
+        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+        return extractXmlTag(response.body(), "BinarySecurityToken");
+    }
+
+    private void signin(String securityToken) throws IOException, InterruptedException {
+        HttpRequest request = HttpRequest.newBuilder()
+                .uri(URI.create(baseUrl + "/_forms/default.aspx?wa=wsignin1.0"))
```

Comment:
That smells the URL that may change without notice 🤪

Reply:
The endpoint to get the security token looks even worse. I didn't find the relevant documentation, you only find that in forums and stackoverflow. Not sure where it comes from in the first place.